src/hotspot/share/opto/loopTransform.cpp

```diff
@@ -1462,113 +1428,43 @@ void PhaseIdealLoop::count_opaque_loop_nodes(Node* n, uint& init, uint& stride)
-          new_init = m->clone();
-          register_new_node(new_init, control);
-        }
-        n->set_req(i, new_init);
+    if (TemplateAssertionPredicateBool::could_be_part(n)) {
+      if (n->Opcode() == Op_OpaqueLoopInit) {
+        init++;
+      } else if (n->Opcode() == Op_OpaqueLoopStride) {
+        stride++;
       } else {
-        assert(op == Op_OpaqueLoopStride, "unexpected opaque node");
-        if (is_unswitched_loop && m->_idx < current && new_stride == nullptr) {
-          new_stride = m->clone();
-          register_new_node(new_stride, control);
-        }
-        if (new_stride != nullptr) {
-          n->set_req(i, new_stride);
-        }
-      }
-      to_clone.set_node(n);
-    }
-    while (true) {
-      Node* cur = to_clone.node();
-      uint j = to_clone.index();
-      if (j+1 < cur->req()) {
-        to_clone.set_index(j+1);
-        break;
-      }
-      to_clone.pop();
-      if (to_clone.size() == 0) {
-        result = cur;
-        break;
-      }
-      Node* next = to_clone.node();
-      j = to_clone.index();
-      if (next->in(j) != cur) {
-        assert(cur->_idx >= current || next->in(j)->Opcode() == Op_Opaque1, "new node or Opaque1 being replaced");
-        if (next->_idx < current) {
-          next = next->clone();
-          register_new_node(next, control);
-          to_clone.set_node(next);
+        for (uint j = 1; j < n->req(); j++) {
+          Node* m = n->in(j);
+          if (m != nullptr) {
+            wq.push(m);
+          }
         }
-        next->set_req(j, cur);
       }
     }
-  } while (result == nullptr);
-  assert(result->_idx >= current, "new node expected");
-  assert(!is_unswitched_loop || new_init != nullptr, "new_init must always be found and cloned");
-  return result;
+  }
 }
 
 // Clone an Assertion Predicate for the main loop. new_init and new_stride are set as new inputs. Since the predicates
 // cannot fail at runtime, Halt nodes are inserted instead of uncommon traps.
 Node* PhaseIdealLoop::clone_assertion_predicate_and_initialize(Node* iff, Node* new_init, Node* new_stride, Node* predicate,
                                                                Node* uncommon_proj, Node* control, IdealLoopTree* outer_loop,
                                                                Node* input_proj) {
-  Node* result = create_bool_from_template_assertion_predicate(iff, new_init, new_stride, control);
+  Node* opaque4_node = iff->in(1);
+  TemplateAssertionPredicateBool template_assertion_predicate_bool(opaque4_node->in(1));
+  BoolNode* new_bool;
+  if (new_stride == nullptr) {
+    new_bool = template_assertion_predicate_bool.clone_and_replace_init(control, new_init, this);
+  } else {
+    new_bool = template_assertion_predicate_bool.clone_and_replace_opaque_loop_nodes(control, new_init, new_stride, this);
+  }
+  opaque4_node = clone_and_register(opaque4_node, control);
+  _igvn.replace_input_of(opaque4_node, 1, new_bool);
```

Comment:
Again, why do we need to separately clone the `opaque4_node`?
And would it not be better to unify the two clone methods, and just pass the `nullptr`, which means clone, and non-nullptr means replace?